Rework history (#88)
silicon-heaven/shv-doc: src/rpcmethods/history.md

before ambiguity may be incorrectly moved by a subsequent time jump and
thus provide inaccurate TimeStamp.

#### Time shifting
A `getLog` implementation must shift recorded times according to time jumps. It

> shouldn't modify times in time jump records. All previous records since the time

Reviewer: > All previous records since the time
jump up to the a time ambiguity record must be considered to be shifted
by a time jump record.

All previous records since the time
jump must be considered to be shifted
by a time jump record always, with or without time ambiguity IMO.

Reviewer 2: That is conditional. In case the time would violate monotonicity, then it must be shifted, but it should not in case it's not. Thus, the statement is correct in the sense that the time jump record should apply to all records until the time jump without ambiguity is located.

> jump up to the a time desynchronization record must be considered to be shifted

Reviewer: ```suggestion
jump up to the a time synchronization record must be considered to be shifted
```
Time jump is time synchronization in my understanding

Author: Should probably say "time ambiguity"

by a time jump record. If there are multiple time jumps in between time
ambiguity records, they are added together.

An optimal implementation of `getLog` for both records and files is to keep an
index of modified times with references to a record ID or a file with an offset
to speed up lookups for `getLog`. Memory constrained devices can implement this
in a less optimal way by keeping only references to the time jumps and
calculating the correct time for every record when requested. The log time
sequence is always kept regardless of date and time and thus this time shifting

> only moves the whole blocks of consistent logs.

Reviewer 2: I am honestly sure if this is something that should be part of the specification. It is written as a dictate for the implementation but in reality it is rather the concluded implementation. It should either be reformulated as a suggested implemntation or it should not he here.

Author: I assume you mean "not sure" 😁

There is also this part:
```
A common detection pattern for time ambiguities is to check the latest record.
If it is in the future, then a time ambiguity event occurred and must be
recorded.
```
Yeah, maybe we should remove most of the implementation tips? Or we could consolidate them in a separate section.

Reviewer 2: Yeah "not sure".

I feel like we should do that. The implementation tips should be seprate. I think that it might be worth it to just split all the "time management" stuff to the separate documentation and it from here because this should be mainly documentation of API. All this stuff muddles it a lot.

I think that we need the document that describes how logs should be recorded so they can be used and exposed by either record or file logs API.

And then there should be a different document that describes how the info provided by file logs and records should be used to assemble the correct wall clock time.

These are two different steps. Possibly the first document is not essentially required but the second one that would deal with the representation of the time and how different real time changes should be handled should be the separate document.

`**/.history/**/.records/*:fetch` or `**/.history/**/.files/*:ls`.

## Time management in logs
Logs are recorded with the device's UTC time. Logs are synchronization from one
instance to another without modification. This means that the date and time of
the logs are always kept as they were on the device that recorded it. This is
ideal when the device has the correct real time clock. When the device does not
have the correct real time clock, time modifications are used to correct it.

There are two types of time modifications recorded in the logs:
- known time jump

> - unknown time desynchronization

Reviewer: ```suggestion
- time ambiguity
```
Both `time ambiguity` and `time desynchronization` are used in the rest of document for this kind of record, I suggest to use the `time ambiguity` only.

`getLog` implementations must shift recorded times according to the time jumps.
More info in the method's description.

> ### Time desynchronization

Reviewer: ```suggestion
### Time ambiguity
```

A time desynchronization can happen only on the first record after boot. If the

> time changes after that, we can calculate a time jump.

Reviewer: ```suggestion
time is synchronized after that, a time jump record is created to enable proper monotonic log time calculation.
```

> A common detection pattern for time desynchronization is to check the latest

Reviewer: this para is not clear to me

Author: It's probably meant to be instructions on when to record the time ambiguities

If it is in the future, then a time ambiguity event occurred and must be
recorded.

> A time ambiguity event creates a break point in the time sequence. Time jumps

Reviewer 3: I would clarify that timejumps that occured AFTER the ambiguity are not applied, but timejumps BEFORE the ambiguity are still applied to the records before ambiguity..

Author: I guess it depends from which side you're going through the records?

Anyway, I think that the point is that a time jump record is only valid between two ambiguities. IOW, A time jump only has an effect on the records that are surrounded by the same two ambiguities as the time jump is.

Reviewer 3: Yeah, it's true before and after may change based on your seek direction.

But having affect on the records surrounded by to ambiguities seems like a good description.

Author: I tried to clarify the thing

are not performed for records before a time desynchronization.

The only exception is in the unlikely event where logs after a time
desynchronization (after all time shifts from jumps are applied) are recorded as
happening before the last log before desynchronization and in such case a time
shift is introduced to move logs before the time desynchronization to be right
before logs after desynchronization. This can really happen only if someone sets
date and time in the future, then powers down the device, resets the RTC and
sets the correct date and time after boot.

### Known time jump

> Known time jump is detected on the device when the system time suddenly doesn't

Reviewer: My understanding is, that the `time jump` record is created, when time is synchronized from external source like GPS or NTP.

Author: I guess that's the usecase, but generally the time can change for whatever reason.

Author: Added a sentence about GPS/NTP

> doesn't correspond to the monotonic time since the last record.

Reviewer 3: ```suggestion
A known time jump is detected on the device when the difference between system time and monotonic time suddenly doesn't correspond to the time difference when the last record was written.
```

Just a suggestion, seems more clear to me.

> Discrepancies of up to 1 second should be disregarded and covered up by time

Reviewer 3: BRCg2 now tweaks time without recording time jump up to 5 seconds difference (mostly because it may be hard to set time precisely when it travels over various networks and then CAN bus). I would let this configurable per device, but in any case anything less than 1 seconds should not be time jump.

Author: Should the doc change then? In my point of view, this paragraph is only an implementation detail, the threshold for time tweaking doesn't actually matter to the API.

Maybe the documentation should be split between what the API should do, and what are the recommended solutions to (time) problems for the devices?

Reviewer 3: Ok, if it's meant just as an implementation detail then it's fine. It just doesn't seem clear to me whether any change more than 1 second should be timejump or not from this sentence.

Author: I think the problem is that for someone consuming this API, this sentence is not relevant. It is a only relevant if you're implementing a device which actually creates these records. I think maybe we should move these instructions for implementing the device into its own section? Or maybe "time tweaking" should be in a different section? If yes, then it could include instructions on what should be a time jump and what shouldn't be a time jump.

But AFAIK, a time jump can be just one second, and it depends on the device, whether it makes a time jump, if it'd be just one second.

So, what this sentence really says is that you cannot make a jump time smaller than 1 second. If you know that your time has a sub-second difference, you should handle it by "time tweaking", for which there unfortunately is no documentation.

What I meant by "implementation detail" is that a device can choose its threshold for a time jump. Though, it can't be less than a second (and for brcg2 it is now 5 seconds as you said).

`**/.history/**/.records/*:fetch` or `**/.history/**/.files/*:ls`.

## Time management in logs

> Logs are recorded with the device's UTC time. Logs are synchronization from one

Reviewer 2: ```suggestion
Logs are recorded with the device's UTC time. Logs are synchronized from one
```

More info in the method's description.

### Time ambiguity event
A time ambiguity event can happen only on the first record after boot. If the
time changes after that, a time jump is calculated instead.

A common detection pattern for time ambiguities is to check the latest record.
If it is in the future, then a time ambiguity event occurred and must be
recorded.

A time ambiguity event creates a break point in the time sequence. Time jumps
only affect those records that are surrounded between the same two time
ambiguities. In other words, a time jump does not have an effect and a record
which is behind a time ambiguity boundary:
```
AMBIGUITY
RECORD <------------------------------------------|
RECORD <------------------------------------------|
TIME-JUMP - this time jump affects these records -|
RECORD <------------------------------------------|
RECORD <------------------------------------------|
AMBIGUITY
RECORD - but not this record
```

The only exception is in the unlikely event where logs after a time ambiguity
event (after all time shifts from jumps are applied) are recorded as happening
before the last log before ambiguity and in such case a time shift is introduced

> to move logs before the time ambiguity event to be right before logs after the

Reviewer 2: Honestly, I know what this means, and I still can't parse the sentence correctly.

How about writing it the other way around?

The only exception is in an unlikely event where logs before ambiguity have a time-jump such that logs would be in the future relative to the logs after the time ambiguity.

Author: I have rewritten this. It is definitely an confusing situation to define. Even @fvacek had problems with understanding it, and we had to make a drawing of it. So maybe the doc should include a drawing?

Also, we are still not sure about the solution for this situation. I think we're going to have to discuss this (the solution) in person, so that we (I and Fanda) have a better understanding of it.

Reviewer 2: If you have a drawing then feel free to include it. One picture is sometimes for thousands words.

We can discuss this in person again...
